DistroLauncher/splash_controller.h

```diff
@@ -0,0 +1,252 @@
+            // temporary and copy it when we can move.
+            ZeroMemory(&procInfo, sizeof(PROCESS_INFORMATION));
+            ZeroMemory(&startInfo, sizeof(STARTUPINFO));
+            startInfo.cb = sizeof(STARTUPINFO);
+            startInfo.hStdInput = stdIn;
+            startInfo.dwFlags |= STARTF_USESTDHANDLES;
+            SetHandleInformation(stdIn, HANDLE_FLAG_INHERIT, 1);
+        }
+        // controller events;
+        struct Events
+        {
+            // Run event transports a non-owning mutating (i.e. borrowing) pointer to the outer controller. Using the
+            // pointer instead of a reference allows copy assignment, which in turn simplifies passing the event object
+            // around by value. With this pointer the Run event will allow the Idle state to access the internals of the
+            // controller to find the pieces required to launch the splash applicaiton, thus avoiding transferring much
+            // heavier pieces of data. Coupling in this case should not be a concern. They are tightly coupled by
+            // definition, since all states and events are members of the outer class, thus all can access even the
+            // controller's private members.
+            struct Run
+            {
+                not_null<SplashController<Strategy>*> controller;
+            };
+            struct ToggleVisibility
+            { };
+
+            struct PlaceBehind
+            {
+                HWND front;
+            };
+
+            struct Close
+            { };
+            using EventVariant = std::variant<ToggleVisibility, Run, PlaceBehind, Close>;
+        };
+
+        // controller states;
+        struct States
+        {
+            // Forward-declaring the states enables declaring the variant upfront because Idle::run() needs the variant
+            // already declared.
+            struct Visible;
+            struct Hidden;
+            struct Idle;
+            struct ShouldBeClosed;
+            using StateVariant = std::variant<Idle, Visible, Hidden, ShouldBeClosed>;
+
+            struct Idle
+            {
+                /// Returns a Visible state if it succeeds in:
+                /// - launching the splash application and
+                /// - finding it's window handle.
+                /// Returns itself (i.e. an instance of the Idle state) otherwise.
+                StateVariant on_event(typename Events::Run event)
+                {
+                    auto controller = event.controller;
+                    if (!Strategy::do_create_process(
+                          controller->exePath, controller->startInfo, controller->procInfo)) {
+                        return *this; // return the the same (Idle) state.
+                    }
+
+                    HWND window = Strategy::do_find_window_by_thread_id(controller->procInfo.dwThreadId);
+                    if (window == nullptr) {
+                        return *this;
```

Comment:
Shouldn’t we log something in those 2 ifs if we failed (for instance, timeout passed and no flutter window visible?)

Reply:
Added some printing for some error situations.

Reply:
This is indeed more informative!